sw/otbn/crypto/field25519.s

```diff
@@ -0,0 +1,350 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/**
+ * This library contains arithmetic for the finite field modulo the prime
+ * p = 2^255-19, which is used for both the X25519 key exchange and the Ed25519
+ * signature scheme.
+ */
+
+
+/**
+ * Fully reduce a 510-bit number modulo p.
+ *
+ * Returns c = a mod p.
+ *
+ * Uses Solinas/pseudo-Mersenne reduction, which is based on the observation
+ * that if a large number x is split into two so that the lower 255 bits are x0
+ * and all bits above 255 are x1, then:
+ *
+ *   x = x0 + (2^255 * x1) \equiv (x0 + (19 * x1)) (mod p)
+ *
+ * If x is large, then x0 + 19 * x1 will be much smaller than x, because 19 <<
+ * 2^255. This step can be repeated as necessary until a conditional
+ * subtraction is enough to fully reduce.
+ *
+ * This routine runs in constant time.
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  w19: constant, w19 = 19
+ * @param[in]  [w21:w20]: a, number to be reduced (a < 2^510)
```

Comment:
Actually, w21 does have the two high bits clear -- this is in fact a saturated representation. Because `bn.mulqacc` sort of really wants you to split at 256 bits, it ended up being easier to use the saturated form, even though unsaturated is more typical for 25519.

Reply:
Right.

Reply:
Although for the first reduction, we actually use `2*p=2^256-38` (since reducing modulo `2*p` is equivalent to reducing modulo p), so in that case `w20=x0` and `w21=x1`. But in terms of the explanation above, if we were just reducing modulo `p`, there's 1 bit of `x1` in `w20`.